### ISSUE TITLE
Cannot exit gracefully while reconnecting

### ISSUE THREAD

**Issue by scop** [gdaws/node-stomp#23]
See https://gist.github.com/scop/e877cc7138884508adca

The problem is that while stompit is reconnecting, it seems that there's no way to exit gracefully by handling SIGINT/SIGTERM. To reproduce, run the script with nothing listening on localhost:61613 and try to exit it with Ctrl-C:

<pre>
Connecting to localhost:61613
Connection error to localhost:61613: connect ECONNREFUSED
Connecting to localhost:61613
Connection error to localhost:61613: connect ECONNREFUSED
[...]
^CShutting down
Channel closed
Connecting to localhost:61613
Connection error to localhost:61613: connect ECONNREFUSED
[...]
</pre>

Looks like there's nothing in stompit that could be called from the shutdown function that would cause the reconnect loop to exit and let the program terminate. After the connection and login has succeeded, the code in the gist seems to work fine, shutdown function gets called and the process exits on Ctrl-C.

**Comment by scop** [comment]
Apparently the same thing also happens in a scenario where a connection + login has succeeded but the channel pool has <code>minFreeChannels</code> > 0, for example if <code>channelFactory</code> in the gist is replaced by:
```javascript
var channelFactory = new stompit.ChannelPool(connections, {
    minFreeChannels: 1,
});
```
...as well as <code>servers.connectHeaders</code> set up so that the connection + login actually succeeds.

**Comment by gdaws** [comment]
Here are the series of commits that fix the problems raised in this issue: 

* 97e2d865c3627114d64b0e3eb444cbb3ef08bb3b Add support for aborting reconnect
* 2a1b05674ba6a46570075d301a2d9b9a6de42dd3 Abort pending connect on channel disconnect
* 84b6d4c8a66c60df78730bc9dd2b0616033eb813 Add ChannelPool#close method
* 1492a457ca13f9ae27ffd4d9d58072598e3b6477 Clear timeout when destroy method is called on a connecting socket
* 3151f42c85408ab83201e08b105afc28a8ee01f7 Destroy transport socket on connect abort

In summary, closing a channel aborts the reconnect occurring in ConnectFailover, and a close method has been added to ChannelPool class that closes any free channels.

In the scenario where you're using a ChannelPool, you need extra signal handlers to shutdown the channel pool.

```js
var closeChannelPool = channelPool.close.bind(channelPool);
process.on("SIGINT", closeChannelPool).on("SIGTERM", closeChannelPool);
```

**Comment by scop** [comment]
Thanks for looking into this, and the fixes.

Maybe I'm missing something wrt the ChannelPool scenario, but code in this gist still does not shut down gracefully on Ctrl-C when there's nothing listening on localhost:61613: https://gist.github.com/scop/aae8d5c3411d1cc92c72

**Comment by gdaws** [comment]
In my test I had signal handlers for the channel pool and the channel in use, and in the nothing-listening scenario the process terminates immediately after SIGINT. Your code needs a signal handler inside the channelPool.channel callback, and it should only exist for as long as the channel is in use.

Closing the channel pool only closes the current free channels and channels that are later reclaimed. Responsibility to close an occupied channel falls on the user. I now realise this behaviour breaks a key feature of the channel pool, is going to make channels awkward to use and it will probably encourage improper use of channel close. Therefore I intend to change channel pool close so that it closes all channels.

In your code there is a call to close the channel in the send callback - this call is unnecessary and is improper use of channel close. Once the send is complete the channel will be automatically reclaimed by the pool. If a channel is used for a subscription then the channel remains occupied until the user calls unsubscribe. A channel is reclaimed by the pool when all sends and subscribes have finished. A channel user shouldn't be concerned with closing the channel or disconnecting the underlying client.

**Comment by choudhary001** [comment]
I have this simple code that I am running in Node.js server but it does not exit once all the messages are sent:

```
var stompit = require('stompit')
var reconnectOptions = { 
    'maxReconnects': 10, 
    'randomize'    : false
};

var connmanager = new stompit.ConnectFailover("failover:(stomp://mqbrokerqa:61613,stomp://failovermqbrokerqa:61613)", reconnectOptions);
connmanager.connect(function(error, client, reconnect) {
    if (error) {
        console.log("terminal error, given up reconnecting: " + error);
        return;
    }   
    client.on('error', function(error) {
        // calling reconnect is optional and you may not want to reconnect if the
        // same error will be repeated.
        //reconnect();
    }); 

   for (i=0; i <= 10; i++) {
       console.log ('sending a message ' + i); 
       client.send('/queue/myqueue', {}, 'Hello number ' + i); 
   }   
   client.send('/queue/myqueue', {}, 'DISCONNECT');
   client.disconnect();
   console.log("Done.");
});

```
Am I using the API incorrectly?
